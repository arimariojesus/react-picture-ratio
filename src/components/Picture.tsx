import React, { CSSProperties, ImgHTMLAttributes } from 'react';
import styles from './Picture.module.css';

import { convertAspectRatioToPercentage as convertRatio } from '../utils/convertAspectRatioToPercentage';

interface PictureProps extends ImgHTMLAttributes<HTMLImageElement> {
  aspectRatio?: string;
  growOnHover?: boolean;
}

const RATIO_DEFAULT = '4:3';

function Picture({
  aspectRatio = RATIO_DEFAULT,
  growOnHover = false,
  ...props
}: PictureProps): JSX.Element {
  const boxStyles = { '--ratio': convertRatio(aspectRatio) } as CSSProperties;

  return (
    <picture
      data-testid="react-picture-ratio"
      className={styles.ratioBox}
      style={boxStyles}
    >
      <img
        src={props.src}
        alt={props.alt}
        loading="lazy"
        data-grow={growOnHover}
        className={styles.image}
      />
    </picture>
  );
}

export default Picture;
