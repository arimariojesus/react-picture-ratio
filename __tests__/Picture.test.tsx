import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Picture from '../src/components/Picture';

describe('<Picture />', () => {
  it('should match snapshot', () => {
    const component = renderer.create(
      <Picture src="https://via.placeholder.com/768x432" alt="Placeholder" />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should be able to show the picture tag and img tag', () => {
    render(
      <Picture src="https://via.placeholder.com/768x432" alt="Placeholder" />,
    );

    const pictureElement = screen.getByTestId('react-picture-ratio');
    expect(pictureElement).toBeInTheDocument();

    const imgElement = screen.getByAltText('Placeholder');
    expect(imgElement).toBeInTheDocument();
  });

  it('picture tag and img tag should have correct className', () => {
    render(
      <Picture src="https://via.placeholder.com/768x432" alt="Placeholder" />,
    );
    const picture = screen.getByTestId('react-picture-ratio');
    const image = picture.firstChild;

    expect(picture).toHaveClass('ratioBox');
    expect(image).toHaveClass('image');
  });

  it('should have shimmer animation', () => {
    render(
      <Picture src="https://via.placeholder.com/768x432" alt="Placeholder" />,
    );
    const picture = screen.getByTestId('react-picture-ratio');

    expect(picture).toHaveClass('shimmerEffect');
  });
});
