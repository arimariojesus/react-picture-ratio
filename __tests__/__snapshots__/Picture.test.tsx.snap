// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Picture /> should match snapshot 1`] = `
<picture
  className=" ratioBox shimmerEffect"
  data-testid="react-picture-ratio"
  style={
    Object {
      "--ratio": "75.00%",
    }
  }
>
  <img
    alt="Placeholder"
    className="image"
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="https://via.placeholder.com/768x432"
  />
</picture>
`;
